docs: Add Committer Expectations document (#52731)
zephyrproject-rtos/zephyr: doc/contribute/committer_expectations.rst

.. _committer-expectations:

Committer Expectations
######################

Overview
********

The Zephyr project encourages you to submit changes as small pull-requests.
Small pull requests (PRs) have the following benefits:

- Reviewed more quickly and reviewed more thoroughly. It's easier for reviewers
  to set aside a few minutes to review small changes several times than it is to

>   allocate a 30 minute block to review a large PR.

Reviewer: It seems minor, but I think it is worth discussing whether this number should remain here.I don't want people to get the mistaken impression that 30 minutes is the worst case scenario for a reviewer. 

In my experience, some reviews take multiple hours or even days if you're going to do them carefully and not just rubber-stamp.

I would suggest something like "allocate a large block of time" for parallelism with the "review a large PR" language.

> - Less wasted work if the PR if reviewers or maintainers reject the direction of

Reviewer: 'if the PR if'?

  the change.

- Easier to rebase and merge. Smaller PRs are less likely to conflict with other
  changes in the tree.

- Easier to rollback if the PR breaks functionality.

Defining Small PRs
==================

- Small PRs should encompass one self-contained logical change.

> - When adding a new feature or API, the PR should address only one part of the

Reviewer 2: ```suggestion
- When adding a new large feature or API, the PR should address only one part of the
```

Some new features may be small and not worth splitting up.

>   feature. In this case create an :ref:`RFC proposal <rfcs>` [#rfc]_ to describe

Reviewer: What is the `[#rfc]_` part doing here?

Reviewer: never mind, resolving this one

  the additional parts of the feature for reviewers.

> - PRs should include tests for any added or changed lines of code.

Reviewer 2: ```suggestion
- PRs should include tests or samples for any added or changed lines of code.
```

Reviewer 3: Is there a particular goal you have in mind by trying to require testing here? Many contributions to Zephyr come from one off small driver additions/features/bugfixes from contributors working on a product. There's already the remarkably high barrier to entry where many PRs die today of simply passing git/code linting steps. Adding yet another brick on the wall feels like we'll lose those first time/small contributors even faster. Sensors are especially prone to this sort of thing. There's over a 100 sensor drivers that more or less are community contributed and tested, with the project providing samples to build against and run.

Speaking of which if a PR contains a bug fix for a piece of hardware what is the real expectation here? If we look at i2c drivers for example very few are actually being tested on hardware given the i2c test case requirement of having a specific external IC connected to it.

Testing by building the driver in a sample or test is about the same, its a smoke test to make sure at least the code builds which is 90% of what we get with CI today when you exclude simulators.

If you consider that, samples and tests are about equivalent in that regard.

Author: IMO, the Zephyr project should move towards 100% test coverage, including drivers.  I think the expectations for tests should be:
1. When fixing a hardware-agnostic bug, PR authors should create a new test that passes with the fix applied, but fails without the fix.
2. When fixing a hardware specific bug (SoC driver, peripheral driver, etc), the PR author should fix the corresponding tests/emulators as well.  If the driver doesn't currently have any tests or emulation, this requirement is waived.  I agree it isn't reasonable to hold up a bug fix and require the PR author to write a new emulator when one doesn't exist.
3. When introducing a new driver, one should also contribute an emulator and tests for the driver.  The PR author has already done a deep dive into the datasheet and has the expertise to write an emulator at the same time.  I fully expect that there will be latent bugs introduced because the PR author writes both the driver and emulator.  However, once the driver is run on real hardware any driver bug fixes will require a fix to the emulator (see previous comment).

- PRs must update any documentation affected by the functional code changes.

- If introducing a new API, the PR must include an example usage of the API.
  This provides context to the reviewer and prevents submitting PRs with unused
  APIs.

Multiple Commits on a Single PR
===============================

> You are further encouraged to break up PR into multiple commits.  Keep in mind

Reviewer: PRs

each commit of your PR should still build cleanly and pass all the CI tests.

For example, when introducing an extension to an API, you can break up the PR
into multiple commits targeting these specific changes:

#. Introduce the new APIs, including shared devicetree bindings
#. Update driver implementation X, with driver specific devicetree bindings
#. Update driver implementation Y
#. Add tests for the new API
#. Add a sample using the API
#. Update the documentation

Large Changes
=============

Large changes to the Zephyr project must submit an :ref:`RFC proposal <rfcs>`
describing the full scope of change and future work.  The :ref:`RFC proposal
<rfcs>` provides the required context to reviewers, but allows for smaller,
incremental, PRs to get reviewed and merged into the project. The RFC should
also define the minimum viable implementation.

Changes which require an :ref:`RFC proposal <rfcs>` include:

- Submitting a new feature.
- Submitting a new API.
- :ref:`treewide-changes`.

> - Other changes which encompass more than 200 lines of non-trivial modifications.

Reviewer: Curious how you arrived at this number as well

PR Requirements
***************

- Each commit in the PR must provide a commit message following the
  :ref:`commit_message_guidelines`.

- All files in the PR must comply with :ref:`Licensing
  Requirements<licensing_requirements>`.

- Follow the Zephyr :ref:`coding_style`.

- PRs must pass all CI checks. This is a requirement to merge the PR, but you
  may mark a PR as draft and explicitly request reviewers to provide early
  feedback.

- When breaking a PR into multiple commits, each commit must build cleanly. This
  is not enforced by the CI system, so it is the PR author’s responsibility to
  verify.

- When major new functionality is added, tests for the new functionality MUST be
  added to the automated test suite. All new APIs MUST be documented and tested

>   and tests MUST cover at least 80% of the added functionality using the code

Reviewer: Do we have some way to measure this? We used to have codecov.io, but that ended up a dud

>   coverage tool and reporting provided by the project..

Reviewer: extra period at end of sentence

- Incompatible changes to APIs must also update the release notes for the

>   next release detailing the change.  APIs marked as experimented are excluded

Reviewer: experimental

  from this requirement. [#api-updates]_

- PRs must also satisfy all :ref:`merge_criteria` before a member of the release
  engineering team will merge the PR into the zephyr tree.

Maintainers may request you break up a PR into smaller PRs and may request you
create an :ref:`RFC proposal <rfcs>`.

Workflow Suggestions That Help Reviewers
========================================

- Don’t resolve comments until the question has been answered or the request

>   completed in an update to the PR.

Reviewer 4: "Unless they applied the reviewer's recommendation exactly, authors must not resolve and hide comments, they must let the initial reviewer do it. The Zephyr project does not require all comments to be resolved. Leaving some completed discussions open can sometimes be useful to understand the greater picture".

- Try to minimize rebases in the middle of a review. If a rebase is required,
  push this as a separate update with no functional changes to the PR. See the
  note below.

- Instead of amending commits, append fixup commits to the PR to address
  reviewer comments. This provides reviewers a clear history of the changes made
  to the PR, along with the rationale for changes. Prior to the final approval
  and merge of the PR, you must perform a squash merge to cleanup the commit

>   history. (TBD - This needs a PoC to demonstrate this workflow).

Reviewer 5: This seems like a particularly good idea, instead of losing history when someone amends the commit and force-pushes.

Reviewer: I'm familiar with `git rebase -i --autosquash` and I use it frequently on my local repositories, but I think this may be unnecessary given the previous bullet. In my experience, separate force-pushes for rebases and code changes show up cleanly enough in the github diffs to make review tolerable.

Reviewer 6: In case anyone missed it before, see https://github.com/zephyrproject-rtos/zephyr-testing/pull/203. Right now, I have a couple fixup commits to respond to review comments, and I'm waiting for people to review those before I squash.

Reviewer Expectations
#####################

- Be respectful when commenting on PRs. Refer to the Zephyr `Code of Conduct`_
  for more details.

- Review PRs within a reasonable time. (TBD - define soft guidelines for

>   “reasonable”)

Reviewer: I continue to want to caution you that there are several areas in the project where review bandwidth is extremely underprovisioned. Big examples are the build system and devicetree bindings. We all want prompt review, but we need to also be realistic about the resources we have as an open source project. I continue to be in favor of prompt review for roadmap items approved by the TSC, but there just isn't enough time in all cases to try to do this project-wide IMO.

Reviewer 6: +1. I think a well understood escalation path is probably more valuable than a very fast review SLO.

Author: Leaving open for discussion by the Process working group.

- Try to provide feedback on the entire PR in one shot. This provides the author

>   an opportunity to address all comments in the next PR update.

Reviewer: Sometimes I see people splitting up reviews in two:

1. first review: nitpicks, code style change requests, etc
2. second: substantive technical comments

This can be useful especially when the nitpick is something like "this PR is doing X inconsistently throughout all its changes and it's making it harder to review; please fix it and I'll take a look at the heart of the changes".

Thoughts on things like that?

Reviewer 6: @mbolivar-nordic: I think the two-stage review you describe is a reasonable review strategy. I think it's very appropriate to send an initial comment to the effect of, "You need to completely restructure this," and then not look at it again until they do to avoid getting into style issues or naming things in code that will go away anyway.

@keith-zephyr: I think reviewing the whole PR is a good way to go from two round trips worth of review to one round trip. For PRs where the reviewer is asking for high-level changes, it seems unlikely that there will just be one round trip, and the reviewer's and author's time might be better spent by getting it right at a high level on one pass before worrying about details. I'm inclined to say this is a judgment call on the part of the reviewer.

Reviewer 6: I don't know what the best way to express this in a document is, but the cost of a partial review in increased round trips goes down when the reviewer is more responsive. To a certain extent, reviewing an entire PR probably increases the round trip time compared to reviewing part of it. I'd rather work with a reviewer who responds one commit or file at a time, every few days, in a large PR than one who avoids the review for weeks at a time, loathing and dreading how long it's going to take them to review the whole thing.

Author: Updating language to call out good examples of partial reviews.

- Partial reviews which focus on a domain (e.g. Devicetree) are permitted, but
  the reviewer must add a comment indicating what portion of the PR they
  reviewed. (TBD - ideally we could designate multiple assignees on a PR to
  cover cross-domain PRs).

- Avoid increasing scope of the PR by requesting new features, especially when
  there is a corresponding :ref:`RFC <rfcs>` associated with the PR. Instead add

>   your suggestions as a comment to the :ref:`RFC <rfcs>`. This also encourages

Reviewer: I think this is an excellent mechanism for avoiding drive-by feature creep requests

Reviewer 7: The opposite may be good: ask to decrease the scope of the PR if it makes sense, maybe split some commit on a separate PR.

Reviewer 6: +1

Author: Regarding decreasing scope, that is mentioned above just below the PR requirements.  Should I move that into the reviewer expectations.

  more collaboration as it is easier for multiple contributors to work on a
  feature once the minimum implementation has merged.

> 

Reviewer 6: @keith-zephyr It might be good to encourage reviewers to call out comments that aren't high-priority for the reviewer and to approve PRs with open comments when they consider the remaining disagreements trivial. (If the author agrees, they can fix it, and the reviewer will presumably approve it again.) That would eliminate a lot of back and forth, and it would decrease the sense of doom on the part of the author when reviewers keep finding new things to comment about.

Reviewer 6: I think so. I would suggest calling that out here and noting that you should "approve" instead of "request changes" if all your remaining comments are non-blocking. Personally, I was surprised by the implications of "request changes" as a new Github user.

- Reviewers should remove themselves from the PR if they are unable to complete
  a review either due to lack of domain knowledge or cannot complete a review in

>   a timely fashion.

Reviewer: And enter into a fight that you can't win against zephyr-bot? I think this is impractical.

Author: Zephyrbot assigns too many reviewers in my opinion.  I was looking for a mechanism for the reviewers to signal to the author (and other reviewers) that they do not plan on looking at a PR.  Removing this language for now.

Only adding a comment stating you don't plan to review doesn't really help move a PR forward.  Github will also continue to spam the uninterested reviewer on any comments or force pushes.

.. _committer-expectations:

Committer Expectations
######################

Overview
********

> The Zephyr project encourages you to submit changes as small pull requests.

Reviewer 7: "Small" may not be the perfect word here but I reckon it's the easiest to pass the idea to an inexperienced user, people that don't do many code reviews have no clear idea of what "easily reviewable" means, and while it's explained right below, I think that the idea of "small" is a good enough starting point for readers that don't get to the details explanation and can't wait to hit the "Open PR" button. But to your point, the kernel documentation does not quite mention "small" in the context of a commit, though it sort-of does in term of a patchseries: https://docs.kernel.org/process/submitting-patches.html#split-changes

> The Zephyr project encourages you to submit changes as small pull requests.

Reviewer 4: > The "Trivial" label is used to mark PRs that are not subject to the 2 day review period.

Found it in doc/project/dev_env_and_tools.rst , thanks.

> The Zephyr project encourages you to submit changes as small pull requests.

Reviewer 4: I just realized most of the text that follows is about "smaller" PRs rather than "small" PRs that we've been struggling to define. This switch would IMHO resolve a few discussions below. In fact the word "smaller" is already used a few times.

```suggestion
The Zephyr project encourages you to submit changes as smaller pull requests.
```

Small pull requests (PRs) have the following benefits:

- Reviewed more quickly and reviewed more thoroughly. It's easier for reviewers
  to set aside a few minutes to review small changes several times than it is to
  allocate large block of time review a large PR.

- Less wasted work if reviewers or maintainers reject the direction of the
  change.

- Easier to rebase and merge. Smaller PRs are less likely to conflict with other
  changes in the tree.

- Easier to revert if the PR breaks functionality.

Defining Small PRs
==================

> - Small PRs should encompass one self-contained logical change.

Reviewer 4: "One self-contained logical change" is already a very common terminology for commits[*]. So this line is practically requesting single-commit PRs. Is that intended?

[*] https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes

  describe the additional parts of the feature for reviewers.

- PRs should include tests for any added or changed lines of code.

- PRs must update any documentation affected by the functional code changes.

- If introducing a new API, the PR must include an example usage of the API.
  This provides context to the reviewer and prevents submitting PRs with unused
  APIs.

Multiple Commits on a Single PR
===============================

You are further encouraged to break up PRs into multiple commits.  Keep in mind
each commit of your PR should still build cleanly and pass all the CI tests.

For example, when introducing an extension to an API, you can break up the PR
into multiple commits targeting these specific changes:

#. Introduce the new APIs, including shared devicetree bindings
#. Update driver implementation X, with driver specific devicetree bindings
#. Update driver implementation Y
#. Add tests for the new API
#. Add a sample using the API
#. Update the documentation

Large Changes
=============

Large changes to the Zephyr project must submit an :ref:`RFC proposal <rfcs>`
describing the full scope of change and future work.  The :ref:`RFC proposal

> <rfcs>` provides the required context to reviewers, but allows for smaller,

Reviewer: Minor: is there a reason for doing two `:ref:`s to the same link in the same paragraph? Seemed a bit odd to me

incremental, PRs to get reviewed and merged into the project. The RFC should
also define the minimum viable implementation.

Changes which require an :ref:`RFC proposal <rfcs>` include:

- Submitting a new feature.
- Submitting a new API.
- :ref:`treewide-changes`.
- Other large changes that can benefit from the :ref:`RFC proposal <rfcs>`
  process. Maintainers have the discretion to request you create an RFC for your
  change.

PR Requirements
***************

- Each commit in the PR must provide a commit message following the
  :ref:`commit_message_guidelines`.

- All files in the PR must comply with :ref:`Licensing
  Requirements<licensing_requirements>`.

- Follow the Zephyr :ref:`coding_style` and :ref:`coding_guidelines`.

- PRs must pass all CI checks. This is a requirement to merge the PR, but you

>   may mark a PR as draft and explicitly request reviewers to provide early

Reviewer 4: > first-time contributors require an org member to approve the CI run for their PRs.

While a little documentation never hurts, does the "user-interface" make this and the next step(s) clear? It really should.

  feedback.

- When breaking a PR into multiple commits, each commit must build cleanly. The
  CI system does not enforce this policy, so it is the PR author's
  responsibility to verify.

- When major new functionality is added, tests for the new functionality shall
  be added to the automated test suite. All API functions should have test cases
  and there should be tests for the behavior contracts of the API. Maintainers
  and reviewers have the discretion to determine if the provided tests are
  sufficient. The examples below demonstrate best practices on how to test APIs
  effectively.

    - `Kernel timer tests`_ provide better than 85% `test coverage`_ for the
      kernel timer.
    - TODO: Provide a driver level API test example.

- Incompatible changes to APIs must also update the release notes for the

>   next release detailing the change.  APIs marked as experimental are excluded

Reviewer: This will be worth highlighting to maintainers; we don't do this now.

create an :ref:`RFC proposal <rfcs>`.

.. _Kernel timer tests: https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/kernel/timer/timer_behavior

.. _test coverage: https://app.codecov.io/gh/zephyrproject-rtos/zephyr/blob/main/kernel/timer.c

Workflow Suggestions That Help Reviewers
========================================

- Don't resolve comments until the question has been answered or the request
  completed in an update to the PR.

- Try to minimize rebases in the middle of a review. If a rebase is required,
  push this as a separate update with no functional changes to the PR. See the
  note below.

- Request re-reviews once comments are addressed. [#re-review]_

- Instead of amending commits, append fixup commits to the PR to address
  reviewer comments. This provides reviewers a clear history of the changes made
  to the PR, along with the rationale for changes. Prior to the final approval
  and merge of the PR, you must perform a squash merge to cleanup the commit
  history. (TBD - This needs a PoC to demonstrate this workflow).

PR Escalation
=============

The Zephyr community is a diverse group of individuals, with different levels of

> commitment and priorities. As such, reviewer and maintainers may not get to your

Reviewer: typo: reviewers

.. _Zephyr Dev Meeting: https://github.com/zephyrproject-rtos/zephyr/wiki/Zephyr-Committee-and-Working-Group-Meetings#zephyr-dev-meeting

.. _Architecture Project: https://github.com/zephyrproject-rtos/zephyr/projects/18

.. _Architecture Working Group: https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group

.. _tsc: https://github.com/zephyrproject-rtos/zephyr/labels/tsc

Reviewer Expectations
#####################

- Be respectful when commenting on PRs. Refer to the Zephyr `Code of Conduct`_
  for more details.

- The Zephyr Project recognizes that reviewers and maintainers have limited
  bandwidth. Prioritize review requests in the following order:

    #. PRs related to items in the `Zephyr Release Plan`_.
    #. PRs you previously reviewed and requested blocking changes.
    #. PRs assigned to you as the area maintainer.
    #. All other PRs.

- Try to provide feedback on the entire PR in one shot. This provides the author
  an opportunity to address all comments in the next PR update.

- Partial reviews are permitted, but the reviewer must add a comment indicating
  what portion of the PR they reviewed. Examples of useful partial reviews

>   include:

Reviewer 4: Not sure what is the best place but I have explained to very many people new to Github that they too can use the "Start Review" button on their _own_ Pull Requests to reduce notification spam by one order of magnitude - and also get a chance to proof-read their comments.

Reviewer 4: Added and commented above, resolving here.

- If introducing a new API, the PR must include an example usage of the API.
  This provides context to the reviewer and prevents submitting PRs with unused
  APIs.

Multiple Commits on a Single PR
===============================

You are further encouraged to break up PRs into multiple commits.  Keep in mind
each commit of your PR must still build cleanly and pass all the CI tests.

For example, when introducing an extension to an API, you can break up the PR
into multiple commits targeting these specific changes:

#. Introduce the new APIs, including shared devicetree bindings
#. Update driver implementation X, with driver specific devicetree bindings
#. Update driver implementation Y
#. Add tests for the new API
#. Add a sample using the API
#. Update the documentation

Large Changes
=============

Large changes to the Zephyr project must submit an :ref:`RFC proposal <rfcs>`
describing the full scope of change and future work.  The RFC proposal provides
the required context to reviewers, but allows for smaller, incremental, PRs to
get reviewed and merged into the project. The RFC should also define the minimum

> viable implementation.

Reviewer 4: "It is possible to share and discuss early code supporting RFCs in large, temporary draft PRs that are  too large to be merged as is but intended to be broken down in smaller PRs later."

Changes which require an RFC proposal include:

- Submitting a new feature.
- Submitting a new API.
- :ref:`treewide-changes`.
- Other large changes that can benefit from the RFC proposal process.
  Maintainers have the discretion to request you create an RFC for your change.

PR Requirements
***************

- Each commit in the PR must provide a commit message following the
  :ref:`commit_message_guidelines`.

- All files in the PR must comply with :ref:`Licensing
  Requirements<licensing_requirements>`.

- Follow the Zephyr :ref:`coding_style` and :ref:`coding_guidelines`.

- PRs must pass all CI checks. This is a requirement to merge the PR. You may
  mark a PR as draft and explicitly request reviewers to provide early feedback,
  even with failing CI checks.

- When breaking a PR into multiple commits, each commit must build cleanly. The
  CI system does not enforce this policy, so it is the PR author's

>   responsibility to verify.

Reviewer: No objections to the content, but I do see some overlap between this and https://docs.zephyrproject.org/latest/contribute/guidelines.html#other-commit-expectations

Author: Provided a link to this content from the general contribution guidelines.

>   responsibility to verify.

Reviewer 7: Would not go as far as suggesting that considering how long each run takes. I understand it's just a suggestion but people unfamiliar with the process would end up doing it all the time and burning half a day on it, it sounds intimidating for no good reason (building for one specific case should be good enough).

>   responsibility to verify.

Reviewer 4: > Would not go as far as suggesting that considering how long each run takes. [...]  and burning half a day on it, 

This is of course done in _parallel_. 

Clarifying:

"In doubt, this verification can be performed by submitting OTHER, even smaller, temporary, test draft PR(s) without any reviewer and closing them unmerged after they passed the CI tests."

> people unfamiliar with the process would end up doing it all the time [...] it sounds intimidating for no good reason 

"In doubt".

In my fairly extensive experience asking people to submit test PRs, including people unfamiliar with github
- they never look spend time having  extra "fun" with github unless they're asked  or unless it makes sense
- when asked, they all immediately  understand that it can and should be done in a different PR.

More generally speaking, _the practice of tests draft PRs should really be encouraged and developed_. There is a different use case where test draft PRs "save lives" and regularly find (or: could have found!) regressions before merge: when running the CI of one git repo on a code change in a git different repo: 

- https://docs.zephyrproject.org/latest/develop/modules.html#process-for-submitting-changes-to-existing-modules
- https://github.com/zephyrproject-rtos/zephyr/pull/53327#issuecomment-1414193515
- https://github.com/thesofproject/rimage/commit/6e22944e05ea7c6d6c2b61e9632bdbf017644323

  effectively.

    - `Kernel timer tests`_ provide better than 85% `test coverage`_ for the
      kernel timer.
    - TODO: Provide a driver level API test example.

- Incompatible changes to APIs must also update the release notes for the
  next release detailing the change.  APIs marked as experimental are excluded
  from this requirement. [#api-updates]_

- Changes to APIs must increment the API version number according to the API
  version rules. [#api-version]_

- PRs must also satisfy all :ref:`merge_criteria` before a member of the release
  engineering team merges the PR into the zephyr tree.

Maintainers may request you break up a PR into smaller PRs and may request you
create an :ref:`RFC proposal <rfcs>`.

.. _Kernel timer tests: https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/kernel/timer/timer_behavior

.. _test coverage: https://app.codecov.io/gh/zephyrproject-rtos/zephyr/blob/main/kernel/timer.c

Workflow Suggestions That Help Reviewers
========================================

- Don't resolve comments until the question has been answered or the request
  completed in an update to the PR.

> - Respond to comments from the "Files changed" view. This allows you to respond

Reviewer 4: ```suggestion
- Respond to comments using the "Start Review" and "Add Review" green buttons in the "Files changed" view. This allows you to respond
```

It's possible to spam one notification per comment (a github bug TBH...) from the "Files Changed" tab.

  to multiple comments and publish the responses in bulk. This reduces the
  number of emails sent to reviewers.

- Try to minimize rebases in the middle of a review. If a rebase is required,

>   push this as a separate update with no functional changes to the PR. When you

Reviewer 6: Non-blocking: I would say, "no other changes since the last push of the PR." Many changes of the type that people ask for in code reviews will not be functional.

  push a rebase only, add a comment to the PR indicating which commit is the
  rebase.

> - Request re-reviews once comments are addressed. [#re-review]_

Reviewer 7: I think users with default permissions can't use this, would not bother with this.

PR Review Escalation
====================

The Zephyr community is a diverse group of individuals, with different levels of
commitment and priorities. As such, reviewers and maintainers may not get to
your PR right away.

The `Zephyr Dev Meeting`_ performs a triage of PRs missing reviewer approval.

> This triage process generally follows this process:

Reviewer: nitpick: this reads a little redundant

#. Identify and update PRs missing an Assignee.
#. Identify PRs without any comments or reviews, ping the PR Assignee to start a
   review or assign to a different maintainer.
#. For PRs that have otherwise stalled, the Zephyr Dev Meeting pings the
   Assignee and any reviewers that have left comments on the PR.

You may escalate your PR outside of the Zephyr Dev Meeting triage process as
follows:

- After a few days of inactivity, ping the Assignee or reviewers on your PR by
  adding a comment to the PR.

- After more than 1 week of inactivity, post a message on the `#pr-help`_
  channel on Discord linking to your PR.

> - After 2 weeks of inactivity, add the `dev-review`_ label to your PR. This

Reviewer: Note: this requires privileges on GitHub that not all contributors have. Might be worth saying something along the lines of "... or ask someone to do it for you if you can't"

  bandwidth. Prioritize review requests in the following order:

    #. PRs related to items in the `Zephyr Release Plan`_.
    #. PRs you previously reviewed and requested blocking changes.
    #. PRs assigned to you as the area maintainer.
    #. All other PRs.

- Try to provide feedback on the entire PR in one shot. This provides the author
  an opportunity to address all comments in the next PR update.

- Partial reviews are permitted, but the reviewer must add a comment indicating
  what portion of the PR they reviewed. Examples of useful partial reviews
  include:

  - Domain specific reviews (e.g. Devicetree).
  - Code style changes that impact the readability of the PR.
  - Reviewing commits separately when the requested changes cascade into the
    later commits.

- Avoid increasing scope of the PR by requesting new features, especially when
  there is a corresponding :ref:`RFC <rfcs>` associated with the PR. Instead add
  your suggestions as a comment to the :ref:`RFC <rfcs>`. This also encourages
  more collaboration as it is easier for multiple contributors to work on a
  feature once the minimum implementation has merged.

> - Mark trivial, non-functional, requests as "Non-blocking" in your comment.

Reviewer 4: ```suggestion
- When using the "Request Changes" option, mark trivial, non-functional, requests as "Non-blocking" in your comment.
```

Otherwise it raises the bar too much for informal discussions, especially for less experienced persons already impressed by public discussions with well known experts.

  Reviewers should approve PRs once only non-blocking changes remain. The PR
  author has discretion as to whether they address all non-blocking comments.

.. _Code of Conduct: https://github.com/zephyrproject-rtos/zephyr/blob/main/CODE_OF_CONDUCT.md

.. _Zephyr Release Plan: https://github.com/orgs/zephyrproject-rtos/projects/13

.. rubric:: Footnotes

.. [#rfc]

   The RFC Proposals doc currently lives under Project Governance. Is there a

>    better location for this doc?

Reviewer: Seems like doc/contribute would be better

.. _committer-expectations:

Committer Expectations
######################

Overview
********

The Zephyr project encourages committers to submit changes as smaller pull
requests. Smaller pull requests (PRs) have the following benefits:

- Reviewed more quickly and reviewed more thoroughly. It's easier for reviewers
  to set aside a few minutes to review smaller changes several times than it is

>   to allocate large block of time review a large PR.

Reviewer 8: ```suggestion
  to allocate large blocks of time to review a large PR.
```

- Less wasted work if reviewers or maintainers reject the direction of the
  change.

- Easier to rebase and merge. Smaller PRs are less likely to conflict with other
  changes in the tree.

- Easier to revert if the PR breaks functionality.

Defining Smaller PRs
====================

- Smaller PRs should encompass one self-contained logical change.

- When adding a new large feature or API, the PR should address only one part of
  the feature. In this case create an :ref:`RFC proposal <rfcs>` to describe the
  additional parts of the feature for reviewers.

- PRs should include tests or samples under the following conditions:

  - Adding new features or functionality.

  - Modifying a feature, especially for API behavior contract changes.

  - Fixing a hardware agnostic bug. The test should fail without the bug fixed
    and pass with the fix applied.

> - PRs must update any documentation affected by the functional code changes.

Reviewer 8: ```suggestion
- PRs must update any documentation affected by a functional code changes.
```

Reviewer 9: `change` not `changes`

- If introducing a new API, the PR must include an example usage of the API.
  This provides context to the reviewer and prevents submitting PRs with unused
  APIs.

Multiple Commits on a Single PR
===============================

Committers are further encouraged to break up PRs into multiple commits.  Keep
in mind each commit in the PR must still build cleanly and pass all the CI
tests.

> For example, when introducing an extension to an API, Committers can break up

Reviewer 8: ```suggestion
For example, when introducing an extension to an API, committers can break up
```

the PR into multiple commits targeting these specific changes:

#. Introduce the new APIs, including shared devicetree bindings
#. Update driver implementation X, with driver specific devicetree bindings
#. Update driver implementation Y
#. Add tests for the new API
#. Add a sample using the API
#. Update the documentation

Large Changes
=============

Large changes to the Zephyr project must submit an :ref:`RFC proposal <rfcs>`
describing the full scope of change and future work.  The RFC proposal provides
the required context to reviewers, but allows for smaller, incremental, PRs to
get reviewed and merged into the project. The RFC should also define the minimum
viable implementation.

Changes which require an RFC proposal include:

- Submitting a new feature.
- Submitting a new API.
- :ref:`treewide-changes`.
- Other large changes that can benefit from the RFC proposal process.

> Maintainers have the discretion to request committers create an RFC for PRs that

Reviewer 6: Non-blocking: "that committers create" or "for committers to create"

I think there are a few other instances of missing prepositions in here.

  be added to the automated test suite. All API functions should have test cases
  and there should be tests for the behavior contracts of the API. Maintainers
  and reviewers have the discretion to determine if the provided tests are
  sufficient. The examples below demonstrate best practices on how to test APIs
  effectively.

    - `Kernel timer tests`_ provide better than 85% `test coverage`_ for the
      kernel timer.
    - Emulators for off-chip peripherals are an effective way to test driver
      APIs. The `smart battery sensor tests`_ use the `smart battery emulator`_,
      providing test coverage for the `fuel gauge API`_ and the `smart battery
      sensor driver`_

- Incompatible changes to APIs must also update the release notes for the
  next release detailing the change.  APIs marked as experimental are excluded
  from this requirement. [#api-updates]_

- Changes to APIs must increment the API version number according to the API
  version rules. [#api-version]_

- PRs must also satisfy all :ref:`merge_criteria` before a member of the release
  engineering team merges the PR into the zephyr tree.

> Maintainers may request committers break up a PR into smaller PRs and may

Reviewer 6: Non-blocking: "Maintainers may request _that_ committers break up a PR into smaller PRs and may request _that_ the commit create an RFC proposal."

> request the committer create an :ref:`RFC proposal <rfcs>`.

Reviewer 8: ```suggestion
request they create an :ref:`RFC proposal <rfcs>`.
```

.. _Kernel timer tests: https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/kernel/timer/timer_behavior

.. _test coverage: https://app.codecov.io/gh/zephyrproject-rtos/zephyr/blob/main/kernel/timer.c

.. _`smart battery sensor tests`: https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/sensor/sbs_gauge/src/test_sbs_gauge.c

> .. _`smart battery emulator`: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/sbs_gauge/emul_sbs_gauge.c

Reviewer: Blocking: please use `:zephyr_file:` to link to files in the tree (grep for examples if unfamiliar). This ensures that release HTML pages point to the corresponding files for that release, unlike direct links, which result in links to files from the future in past release HTML.

.. _`fuel gauge API`: https://app.codecov.io/gh/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/fuel_gauge.h

.. _`smart battery sensor driver`: https://app.codecov.io/gh/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/sbs_gauge/sbs_gauge.c

> 

Reviewer: I'm unsure if these codecov results are usable here for a similar reason.

Workflow Suggestions That Help Reviewers
========================================

- Unless they applied the reviewer's recommendation exactly, authors must not
  resolve and hide comments, they must let the initial reviewer do it. The
  Zephyr project does not require all comments to be resolved before merge.
  Leaving some completed discussions open can sometimes be useful to understand
  the greater picture.

- Respond to comments using the "Start Review" and "Add Review" green buttons in

>   the "Files changed" view. This allows you to respond to multiple comments and

Reviewer 8: This is the only occurence of "you" in the whole document. At the risk of sounding nitpicky, I'd just remove it from the sentence?

```suggestion
  the "Files changed" view. This allows to respond to multiple comments and
```
